### PR TITLE
Make dotnet-pgo capable of handling dll conflicts by providing deconfliction options

### PR DESCRIPTION
**Pull request by davidwrighton**
Now --reference will specify a prefferred dll, or if the developer is unable to specify a preferred dll, they can disable all automatic reference loading and specify the versions of all other dlls to be used.

### CODE REVIEW

File: src/coreclr/tools/dotnet-pgo/Program.cs [hunk 1]
```diff
@@ -1181,6 +1184,64 @@ static int InnerProcessTraceFileMain(CommandLineOptions commandLineOptions)
...
+                    var managedModule = module.ManagedModule;
+
+                    if (module.ClrInstanceID != clrInstanceId.Value)
+                        continue;
+
+                    if (managedModule.ModuleFile != null)
+                    {
+                        string simpleName = managedModule.ModuleFile.Name;
+                        if (simpleName.EndsWith(".il"))
+                            simpleName = simpleName.Substring(0, simpleName.Length - 3);
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/Program.cs:1199]:
```suggestion
                            simpleName = simpleName[..^3];
```

File: src/coreclr/tools/dotnet-pgo/Program.cs [hunk 2]
```diff
@@ -1181,6 +1184,64 @@ static int InnerProcessTraceFileMain(CommandLineOptions commandLineOptions)
...
+
+                        // This path may be normalized
+                        if (File.Exists(filePathTemp) || !tsc._normalizedFilePathToFilePath.TryGetValue(filePathTemp, out candidateFilePath))
+                            candidateFilePath = filePathTemp;
+
+                        if (!duplicateModuleAnalysis.TryGetValue(simpleName, out HashSet<string> candidatePaths))
+                        {
+                            duplicateModuleAnalysis[simpleName] = candidatePaths = new HashSet<string>();
+                        }
+                        duplicateModuleAnalysis[simpleName].Add(candidateFilePath);
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/Program.cs:1212]:
```suggestion
                        candidatePaths.Add(candidateFilePath);
```

File: src/coreclr/tools/dotnet-pgo/Program.cs [hunk 3]
```diff
@@ -1181,6 +1184,64 @@ static int InnerProcessTraceFileMain(CommandLineOptions commandLineOptions)
...
+                bool duplicateError = false;
+                foreach (var assembliesWithDuplicates in duplicateModuleAnalysis)
+                {
+                    if (assembliesWithDuplicates.Value.Count == 1)
+                        continue;
+
+                    ModuleDesc loadedViaReference = null;
+                    foreach (var module in modulesLoadedViaReference)
+                    {
+                        if (String.Compare(module.Assembly.GetName().Name, assembliesWithDuplicates.Key, StringComparison.OrdinalIgnoreCase) == 0)
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/Program.cs:1225]:
```suggestion
                        if (string.Equals(module.Assembly.GetName().Name, assembliesWithDuplicates.Key, StringComparison.OrdinalIgnoreCase))
```

File: src/coreclr/tools/dotnet-pgo/Program.cs [hunk 4]
```diff
@@ -1210,9 +1272,19 @@ static int InnerProcessTraceFileMain(CommandLineOptions commandLineOptions)
...
                             {
-                                PrintError($"Dll mismatch between assembly located at \"{e.ModuleILPath}\" during trace collection and module \"{tsc.PEReaderToFilePath(ecmaModule.PEReader)}\"");
-                                mismatchErrors++;
-                                mismatch = true;
+                                if (modulesLoadedViaReference.Contains(ecmaModule) && duplicateModuleAnalysis[ecmaModule.Assembly.GetName().Name].Count > 1)
+                                {
+                                    // This is the case where a duplicate dll mismatch was avoided by specifying a -reference parameter
+                                    PrintMessage($"Disabling load of assembly data from assembly located at \"{e.ModuleILPath}\" during trace collection as module \"{tsc.PEReaderToFilePath(ecmaModule.PEReader)}\" is preferred, and does not match");
+                                    idParser.RemoveModuleIDFromLoader(e.ModuleID);
+                                    mismatchHandled = true;
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/Program.cs:1280]: So essentially this allows dotnet-pgo to work even if multiple assemblies with the same name were loaded, by specifying one of them explicitly in `--reference`?

**Review by jakobbotsch — APPROVED** [review]
LGTM, a few nits and a question

File: src/coreclr/tools/dotnet-pgo/Program.cs [hunk 5]
```diff
@@ -1181,6 +1184,64 @@ static int InnerProcessTraceFileMain(CommandLineOptions commandLineOptions)
                 if (!tsc.Initialize())
                     return -12;
 
+                Dictionary<string, HashSet<string>> duplicateModuleAnalysis = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/tools/dotnet-pgo/Program.cs:1187]: The way etl trace processessing works will forcibly lowercase stuff very often. I'm not pleased about it, but if we want most things to work, they need to be handled with logic that ignores casing.

File: src/coreclr/tools/dotnet-pgo/Program.cs (hunk 4 above)
Comment (davidwrighton) [reply src/coreclr/tools/dotnet-pgo/Program.cs:1280]: Yes. Its not ideal, and results in a giant pile of warnings about being unable load load types, but it does work.

**Pull request merged**